drivers/meter/ade7913/ade7913.c

```diff
@@ -0,0 +1,2308 @@
+	no_os_free(dev);
+
+	return ret;
+}
+
+/**
+ * @brief Remove the device and release resources.
+ * @param dev - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade7913_remove(struct ade7913_dev *dev)
+{
+	int ret;
+
+	ret = no_os_spi_remove(dev->spi_desc);
+	if (ret)
+		return ret;
+
+	no_os_free(dev);
+
+	return 0;
+}
+
+/**
+ * @brief Reset the device using SW reset.
+ * @param dev - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade7913_sw_reset(struct ade7913_dev *dev)
+{
+	int ret;
+	uint8_t ver_product;
+
+	ret = ade7913_write(dev, ADE7913_REG_SWRST, ADE7913_SWRST_CMD, ADE7913_S_OP);
+
+	if (ret)
+		return ret;
+
+	dev->crc_en = 1;
+
+	/* Wait for device to initialize */ // ToDo - check required timing
+	no_os_mdelay(100);
+
+	/* Read version product */
+	ret = ade7913_get_version_product(dev, &ver_product);
+	if (ret)
+		return ret;
+
+	dev->ver_product = ver_product;
+
+	return 0;
+}
+
+/**
+ * @brief Reset the device using SW reset.
+ * @param dev - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade7913_hw_reset(struct ade7913_dev *dev)
+{
+	int ret;
+	uint8_t ver_product;
+
+	ret = no_os_gpio_set_value(dev->gpio_reset, NO_OS_GPIO_LOW);
+	if (ret)
+		return ret;
+
+	no_os_mdelay(10);
+
+	ret = no_os_gpio_set_value(dev->gpio_reset, NO_OS_GPIO_HIGH);
+	if (ret)
+		return ret;
+
+	dev->crc_en = 1;
+
+	/* Wait for device to initialize */ // ToDo - check required timing
+	no_os_mdelay(100);
+
+	/* Read version product */
+	ret = ade7913_get_version_product(dev, &ver_product);
+	if (ret)
+		return ret;
+
+	dev->ver_product = ver_product;
+
+	return 0;
+}
+
+/**
+ * @brief Convert a 24-bit raw sample to millivolts.
+ * @param dev - The device structure.
+ * @param ch - Device channel.
+ * @param mv_val - Value in millivolts.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade7913_convert_to_millivolts(struct ade7913_dev *dev,
+				  enum ade7913_wav_e ch, int32_t *mv_val)
+{
+	int64_t value = 0;
+
+	if (!dev)
+		return	-ENODEV;
+
+	if (ch > ADE7913_V2_WAV)
+		return	-EINVAL;
+
+	switch (ch) {
+	case ADE7913_I_WAV:
+		/* times 2, two's complement data */
+		// Scale is 0,03125 = 3125/100000 = 1/32 = 1/2^5
+		value = ((int64_t)ADE7913_VREF * dev->i_wav) / (1 << 28);
+		break;
+	case ADE7913_V1_WAV:
+		value = ((int64_t)ADE7913_VREF * dev->v1_wav) / (1 << 23);
+		break;
+	default:
+		value = ((int64_t)ADE7913_VREF * dev->v2_wav) / (1 << 23);
+		break;
+	}
+
+	*mv_val = (int32_t)value;
+
+	return 0;
+}
+
+/**
+ * @brief Get STREAM_DBG mode.
+ * @param dev - The device structure.
+ * @param stream_dbg - Read debug mode setting read.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade7913_get_stream_dbg_mode(struct ade7913_dev *dev,
+				enum ade7913_stream_debug_e *stream_dbg)
+{
+	int ret;
+	uint8_t reg_val;
+
+	if (!dev)
+		return	-ENODEV;
```

Comment:
not sure if this check is needed, I would actually remove from all APIs which call in the end SPI read API because the SPI read API already checks for this. it would be better to check if stream_dbg is null

Reply:
Check removed from the functions that call the SPI read API.